enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -0,0 +1,809 @@
+runs is deleted.
+
+Different
+[DeletionPropagation](https://github.com/kubernetes/apimachinery/blob/3e2600dc79feea6cdc8a9224bc8a6a7fcfee1466/pkg/apis/meta/v1/types.go#L526)
+values will impact how Kubernetes will garbage collect the `PersistentIP`s when
+deleting the VM object from the datastore. The default deletion propagation is
+`background`.
+
+#### Deleting the VM with background `DeletionPropagation`
+
+The following sequence diagram describes how Kubernetes GC works for the
+KubeVirt entities (VM / VMI / pod):
+
+```mermaid
+sequenceDiagram
+  actor User
+  participant persistentIP
+  participant GC
+  participant VM
+  participant VMI
+  participant launcherPod
+
+  User->>VM: Delete VM
+  VM->>VMI: Delete VMI
+  activate VM
+  VMI->>launcherPod: Delete launcherPod
+  activate VMI
+  launcherPod-->>VMI: launcherPod Deleted
+  Deactivate VMI
+  note over VMI: clear VMI finalizer
+  VMI-->>VM: VMI Deleted
+  Deactivate VM
+  note over VM: clear VM finalizer
+  VM-->>User: VM Deleted
+  VM->>GC: VM deleted
+  GC->>persistentIP: Delete persistentIP
+  persistentIP-->>GC: persistentIP Deleted
+```
+
+The two finalizers mentioned above ensure the VM object is only removed from
+the datastore after the VM has actually stopped. Since the `PersistentIP`
+allocation is only removed **after** the VM is gone, we have avoided the
+potential issue previously described.
+
+#### Deleting the VM with foreground `DeletionPropagation`
+
+When using the foreground `DeletionPropagation`, the dependent object's - e.g.
+the VMI, and the persistentIP allocations - deletion will be triggered
+**before** the VM is actually deleted.
+
+Given the existing finalizers set in the VM and VMI object, the VM object will
+not be deleted while the pod is in the datastore - but the corresponding
+`PersistentIP` CR will be scheduled for deletion.
+
+Thus, when the foreground deletion propagation is chosen, the IP is returned to
+the pool while the VM is still shutting down (or about to start to ...), which
+will lead to the probable scenario described above: another workload will be
+assigned the IP for a VM which is still running.
+
+#### Deleting the VM with orphaned `DeletionPropagation`
+
+When using the orphaned `DeletionPropagation`, the dependent objects - e.g.
+will **not** be deleted by the garbage collector. I.e. the user is asking for
+the dependent resources to leak.
+
+#### Recommendation
+
+The simplest way to work-around this issue is simply to only allow `background`
+propagation deletions.
+
+Our proposal is for the KubeVirt validating webhook to reject VM foreground/
+orphaned delete propagations **whenever** it features an interface for which
+a persistent IP was requested.
+
+#### Alternatives for preventing the premature release of an IP to the pool
+
+We could also decorate the `PersistentIP` with a finalizer, and remove it once
+the corresponding VMI (or pod ...) is deleted. This would require the
```

Comment:
upon sync, it should remove finalizers from `PersistentIP` that don't have a matching pod,
for example if DEL event is lost.

Can DEL event lost without a sync happening (i.e CNI allocated PIP, added finalizers, pod failed to be created, DEL event lost, but not sync happens because control pod didn't crash) ?
In this case we would need a pool repair controller

Reply:
Pod teardown is triggered on the ovnkube-control plane when the pod status changes to succeeded or failed, or on pod deletion. 

Thus, it doesn't matter if the CNI DEL event is missed.